### ISSUE TITLE
We need Path MTU discovery

### ISSUE THREAD

**Issue by IntegratedQuantum** [PixelGuys/Cubyz#819]
On some connections the MTU is less than 1500 bytes, which may silently drop packets above the real MTU.
Always setting the MTU to the IPv4 minimum of 576 bytes is not a good solution, as it slows down everyone.
It might be worth looking into RFC 8899 (Datagram Packetization Layer Path MTU Discovery) which is used by QUIC.

**Comment by alexdesander** [comment]
I sadly can't contribute code because I don't know zig well, have uni stress and need to work on my own projects, but here's an idea how you could fix it temporarly in a better way:

The client dictates MTU size. Send client hellos from very large to very low. Every time a client hello is sent and no server hello returned, make the client hello smaller and repeat.

The server retrieves the mtu size from the size of the client hello and adds that information to a connection and assembles packets with that MTU in mind. Now the connection has a fixed MTU, but one that works (most likely).

I think this is much easier to implement than a real mtu discovery algorithm. Sadly mtu discovery is a chaotic topic because the internet is basically the wild west.

**Comment by IntegratedQuantum** [comment]
I'll consider this if it takes too long to implement real MTU.

**Comment by Wunka** [comment]
RFC 8899 describes these 3 different probe message variants:
> Probing using padding data:
A probe packet that contains only control information together with any padding, which is needed to inflate to the size of the probe packet. Since these probe packets do not carry an application-supplied data block, they do not typically require retransmission, although they do still consume network capacity and incur endpoint processing.

>Probing using application data and padding data:
A probe packet that contains a data block supplied by an application that is combined with padding to inflate the length of the datagram to the size of the probe packet.

>Probing using application data:
A probe packet that contains a data block supplied by an application that matches the size of the probe packet. This method requests the application to issue a data block of the desired probe size.

The first one would probably involve making an extra channel for it. 
If we use the second or third we would have to decide which channel will to that because we will expect some missing packets when the probe size is to big. (but not much as we will not send that many and often probe messages)
The first one would probably make the logic a bit cleaner. 

From what I have read our network implemention only lacks these probe messages which will then change the `mtuEstimate` variable. For debugging purposes I would also say we add the `mtuEstimate` to the advanced network debug window

**Comment by IntegratedQuantum** [comment]
When I looked into it I think one of the main issues was how to detect and react to MTU reductions (which could theoretically happen if your traffic is rerouted for some reason)

> If we use the second or third we would have to decide which channel will to that because we will expect some missing packets when the probe size is to big.

The obvious answer is the slow protocol, the data inside is supposed to assume that it's fine to arrive late.

Overall which method we choose probably doesn't matter, we just need to ensure that the math is right and that the cost of the probe packet is always significantly less than the expected gains.

In the end it's probably just a question of which is easier to implement.

**Comment by Wunka** [comment]
I looked at his a little bit and sadly on the `/server restart` even uses the slow channel... so maybe until #1319 is done, this is sadly not nice testable / usable via the slow channel.

**Comment by IntegratedQuantum** [comment]
What does server restart have to do with MTU discovery?

**Comment by Wunka** [comment]
Its the only thing which uses the slow channel. (which we want to use for mtu probe messages)

**Comment by IntegratedQuantum** [comment]
Why should it even use a channel? The channels are meant for ensuring that packets reach their target, MTU probe is a one-shot packet

**Comment by Wunka** [comment]
As said in my other comment: https://github.com/PixelGuys/Cubyz/issues/819#issuecomment-4924904550 there are 3 ways to do probe messages. 
For me it seemed also after your comment that the slow channel would maybe be the correct point to use way 2 or 3. 
Of course we can also use 1.

**Comment by IntegratedQuantum** [comment]
Oh right, now I get it. Can't you just send an empty packet if no data is available?

**Comment by Wunka** [comment]
Yeah, I think I will just still use the slow channel and then if I want to send probe messages still send messages even if there is nothing to send
